src/nectarchain/calibration/NectarGain/SPEfit/utils.py

```diff
@@ -0,0 +1,159 @@
+import logging
+import math
+
+import numpy as np
+from iminuit import Minuit
+from scipy import interpolate, signal
+from scipy.special import gammainc
+
+logging.basicConfig(format='%(asctime)s %(name)s %(levelname)s %(message)s')
+log = logging.getLogger(__name__)
+log.handlers = logging.getLogger('__main__').handlers
+
+__all__ = ['UtilsMinuit','Multiprocessing']
+from .parameters import Parameters
+
+class Multiprocessing() : 
+    @staticmethod
+    def custom_error_callback(error):
+        log.error(f'Got an error: {error}')
+        log.error(error,exc_info=True)
+
+class UtilsMinuit() :
+    @staticmethod
+    def make_minuit_par_kwargs(parameters : Parameters):
+        """Create *Parameter Keyword Arguments* for the `Minuit` constructor.
+
+        updated for Minuit >2.0
+        """
+        names = parameters.parnames
+        kwargs = {"names": names,"values" : {}}
+
+        for parameter in parameters.parameters:
+            kwargs["values"][parameter.name] = parameter.value
+            min_ = None if np.isnan(parameter.min) else parameter.min
+            max_ = None if np.isnan(parameter.max) else parameter.max
+            error = 0.1 if np.isnan(parameter.error) else parameter.error
+            kwargs[f"limit_{parameter.name}"] = (min_, max_)
+            kwargs[f"error_{parameter.name}"] = error
+        return kwargs
+    
+    @staticmethod
+    def set_minuit_parameters_limits_and_errors(m : Minuit,parameters : dict) :
+        """function to set minuit parameter limits and errors with Minuit >2.0
+
+        Args:
+            m (Minuit): a Minuit instance
+            parameters (dict): dict containing parameters names, limits errors and values
+        """
+        for name in parameters["names"] :
+            m.limits[name] = parameters[f"limit_{name}"]
+            m.errors[name] = parameters[f"error_{name}"]
+
+
+# Usefull fucntions for the fit
+def gaussian(x, mu, sig):
+    return (1./(sig*np.sqrt(2*math.pi)))*np.exp(-np.power(x - mu, 2.) / (2 * np.power(sig, 2.)))
+
+def weight_gaussian(x,N, mu, sig) : 
+    return N * gaussian(x, mu, sig)
+
+def doubleGauss(x,sig1,mu2,sig2,p):
+    return p *2 *gaussian(x, 0, sig1) + (1-p) * gaussian(x, mu2, sig2)
+
+def PMax(r):
+    if (np.pi*r**2/(np.pi*r**2 + np.pi - 2*r**2 - 2) <= 1):
+        return np.pi*r**2/(np.pi*r**2 + np.pi - 2*r**2 - 2)
+    else:
+        return 1
+
+def ax(p,res):
+    return ((2/np.pi)*p**2-p/(res**2+1))
+
+def bx(p,mu2):
+    return (np.sqrt(2/np.pi)*2*p*(1-p)*mu2)
+
+def cx(sig2,mu2,res,p):
+    return (1-p)**2*mu2**2 - (1-p)*(sig2**2+mu2**2)/(res**2+1)
```

Comment:
I detected a typo error here, the equation (4) of Caroff et al. 2019 was not well written

Reply:
Indeed. For the record here, @guillaumegrolleron and I discussed this and saw an inconsistency (presumably a typo in the paper) between Equation 4 in [Caroff et al. (2019)](https://ui.adsabs.harvard.edu/abs/2019SPIE11119E..1WC/abstract) and Sami's original implementation in `calin` (see https://github.com/llr-cta/calin/blob/main/python/SPEFit.py).

Reply:
Perhaps this typo should be documented in the code so that future reader/you would not correct in the other way to match the paper :-)

Reply:
Pierre Jean confirms the typos, which was earlier confirmed by Sami in 2019.